mgr/dashboard: Add support for RBD Trash

tspmelo opened the pull request:
This PR will bring all RBD Trash operations to the ceph dashboard.

Fixes: http://tracker.ceph.com/issues/24272

There is a known issue with the date picker component of `ngx-bootstrap`.
There is already a PR to fix this, so I will wait for the next release and then update this PR.

- [x] ~~Update `ngx-bootstrap`~~
- [x] Add expiration date validation
- [x] Add i18n in all strings

tspmelo commented:
Screenshots:
![image](https://user-images.githubusercontent.com/399326/43469820-f6a8feea-94de-11e8-9d7f-e18e83889e57.png)
![image](https://user-images.githubusercontent.com/399326/43469846-073b53fc-94df-11e8-954c-40a3890a510c.png)
![image](https://user-images.githubusercontent.com/399326/43469883-1630abdc-94df-11e8-860e-92ee360b4b9b.png)
![image](https://user-images.githubusercontent.com/399326/43469894-1e4667b2-94df-11e8-9afc-149e983a46c1.png)
![image](https://user-images.githubusercontent.com/399326/43469911-289c9380-94df-11e8-88c8-36b36ce7aee1.png)
![image](https://user-images.githubusercontent.com/399326/43469921-335e2d38-94df-11e8-81cc-f466692a199b.png)
![image](https://user-images.githubusercontent.com/399326/43469930-3a30f654-94df-11e8-83a8-ba560972c642.png)

ricardoasmarques commented:
This is the first page with top-level tabs, but in the future we will have more pages like this (grafana integration, etc...), so I suggest to reflect the selected tab in the URL and breadcrumb.

Some benefits: bookmark selected tab, browser back navigation, use router to navigate directly to a tab, etc...

tchaikov commented:
![](https://user-images.githubusercontent.com/399326/43469846-073b53fc-94df-11e8-954c-40a3890a510c.png)

shall we use "Move an image to trash" instead of "Move RBD to trash"? it sounds like this button will help the user to disable RBD as a whole.

tspmelo commented:
Should we use "RBD Image" or is "Image" enough?

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+                result.append(trash)
+            return result
+
+    def _trash_list(self, pool_name=None):
```
Although they were identical, the _rbd_pool_list call was referencing different methods.
I have renamed _rbd_pool_list  to _trash_pool_list, so we can better differentiate them.

Refactoring this would require I added parameters to the method and conditional calls inside, do you think it is worth it?

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+            result = []
+            for trash in images:
+                trash['pool_name'] = pool_name
+                trash['deletion_time'] = "{}Z".format(trash['deletion_time'].isoformat())
```
ceph cmds/api doesn't return dates in iso format, which can cause problems in the UI.

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -374,6 +375,17 @@ def default_features(self):
         rbd_default_features = mgr.get('config')['rbd_default_features']
         return _format_bitmask(int(rbd_default_features))
 
+    # Move an image to the trash. Images, even ones actively in-use by clones,
```
Done.

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -5,17 +5,18 @@
 
-import cherrypy
 import six
+import cherrypy
```
I'm guessing it was during a rebase.

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+
+        for pool in images:
+            for image in pool['value']:
+                if image['deferment_end_time'] < now:
```
They are both in UTC.
rbd api doesn't support the purge command, so I implemented it. 
When purging you can only delete images that have expired, and that is determined by the `deferment_end_time`

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+    @RESTController.Collection('POST', query_params=['pool_name'])
+    def purge(self, pool_name=None):
+        now = "{}Z".format(datetime.now().isoformat())
+        images = self._trash_list(pool_name)
```
done.

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+                if image['deferment_end_time'] < now:
+                    def _remove_trash(ioctx):
+                        rbd_inst = rbd.RBD()
+                        # pylint: disable=cell-var-from-loop
```
done

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+    @handle_rados_error('pool')
+    @RbdTask('trash/purge', ['{pool_name}'], 2.0)
+    @RESTController.Collection('POST', query_params=['pool_name'])
+    def purge(self, pool_name=None):
```
`purge` is the correct name for this method, but I have renamed `delete` into `remove`.

tspmelo replied:
It's the official command to execute this operation.
`trash purge [pool-name]`

sebastian-philipp replied:
urgh. thus there is an inconsistency between the command line and the python API?

tspmelo replied:
`Remove` is also a valid rbd trash command, so the API is correct.
What happens is that the python API is lacking the purge method, so I implemented it in the dashboard.

tspmelo commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -453,3 +465,81 @@ def _clone(ioctx):
+
+        return _rbd_call(pool_name, _restore_trash)
+
+    # Delete an image from trash. If image deferment time has not expired you
```
done

tspmelo commented:
@dillaman 

1. Done.
2. The message informs if the command was successfully executed. It doesn't know if there was any image deleted. Was your image expired? could you try again with the new code?
3. Done.
4. I have removed the need to type the image name when deleting an expired image. CLI doesn't require you to do it, so the dashboard should do the same.
5. I have change the column from "expires at" to "status" and I'm using the same output the CLI returns. Now it will show "Expired at 8/22/18 4:16:29 PM" or "Protected until 8/31/18 4:16:11 PM"

tspmelo commented:
I have updated the screenshots: https://github.com/ceph/ceph/pull/23351#issuecomment-409265416

votdev commented:
![44475947-ee7d7400-a625-11e8-88a4-044d12a45ffa](https://user-images.githubusercontent.com/1897962/44530562-f81ede80-a6ee-11e8-9d37-53f24708197c.png)

Require the user to enter this long text to confirm the deletion is a really complicated and annoying workflow on touch devices (and also on desktops). Wouldn't it better to force a 'Yes' and add a 'Force' checkbox?

votdev commented:
![auswahl_004](https://user-images.githubusercontent.com/1897962/44531424-0c63db00-a6f1-11e8-9bb3-b8ef394d816b.png)

The columns ``Status`` and ``Deleted At`` should be left aligned.

tspmelo commented:
@votdev 
Since I removed the need to confirm if the image has expired, I could remove the --force.
But writing the name of the image is what we do in all the deletion modals, so its better to leave it like this.
If we decide to change this, it should be a separate PR.

I still need to add validation to the date field. Past dates can be one of them.

votdev commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -15,7 +16,7 @@
 from ..security import Scope
 from ..services.ceph_service import CephService
-from ..tools import ViewCache
+from ..tools import ViewCache, str_to_bool
```
Use ``from distutils.util import strtobool`` if possible.

tspmelo replied:
if `strtobool` does everything `str_to_bool` does, then we should replace it everywhere.

tspmelo replied:
I will have a look

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html:
```diff
@@ -0,0 +1,45 @@
+        </p>
+
+        <div class="form-group">
+          <label for="expires">Protection expires at</label>
```
Missing i18n

tspmelo replied:
done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html:
```diff
@@ -0,0 +1,45 @@
+          [formGroup]="moveForm"
+          novalidate>
+      <div class="modal-body">
+        <p>
```
Whole text is not translated.

tspmelo replied:
done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-purge-modal/rbd-trash-purge-modal.component.html:
```diff
@@ -0,0 +1,50 @@
+      <div class="modal-body">
+        <p>
+          <ng-container i18n>To purge, select one or All images and click on</ng-container>&nbsp;
+          <kbd>Purge Trash</kbd>.&nbsp;
```
i18n

tspmelo replied:
done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-purge-modal/rbd-trash-purge-modal.component.html:
```diff
@@ -0,0 +1,50 @@
+          <label class="center-block">Pool:
+            <input class="form-control"
+                   type="text"
+                   placeholder="Pool name..."
```
Placeholder is not translated

tspmelo replied:
done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html:
```diff
@@ -0,0 +1,53 @@
+          novalidate>
+      <div class="modal-body">
+        <p>
+          To restore
```
Text is not translated

tspmelo replied:
done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html:
```diff
@@ -0,0 +1,53 @@
+                 id="name"
+                 autocomplete="off"
+                 formControlName="name"
+                 placeholder="image_name"
```
Is this the correct placeholder text? Don't forget i18n.

tspmelo replied:
Removed the placeholder.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html:
```diff
@@ -0,0 +1,45 @@
+        <div class="form-group">
+          <label for="expires">Protection expires at</label>
+          <input type="text"
+                 placeholder="NOT PROTECTED"
```
Placeholder is not translated

tspmelo replied:
Done.

LenzGr commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-delete-modal/rbd-trash-delete-modal.component.html:
```diff
@@ -0,0 +1,75 @@
+          <ng-container *ngIf="pattern === ''">
+            <ng-container i18n>To confirm the deletion of</ng-container>&nbsp;
+            <kbd>{{ poolName }}/{{ imageName }}@{{ imageId }}</kbd>&nbsp;
+            <ng-container i18n>click on</ng-container>&nbsp;
```
s/click on/click/

LenzGr commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html:
```diff
@@ -0,0 +1,48 @@
+      <div class="modal-body">
+        <p>
+          <ng-container i18n>To move</ng-container>
+          <kbd>{{ poolName }}/{{ imageName }}</kbd> to trash, click on
```
s/click on/click/

LenzGr commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-purge-modal/rbd-trash-purge-modal.component.html:
```diff
@@ -0,0 +1,51 @@
+          novalidate>
+      <div class="modal-body">
+        <p>
+          <ng-container i18n>To purge, select one or All images and click on</ng-container>&nbsp;
```
s/click on/click/

LenzGr commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html:
```diff
@@ -0,0 +1,53 @@
+      <div class="modal-body">
+        <p>
+          To restore
+          <kbd>{{ poolName }}/{{ imageName }}@{{ imageId }}</kbd>, type the image's new name and click on
```
s/click on/click/

LenzGr commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-delete-modal/rbd-trash-delete-modal.component.html:
```diff
@@ -0,0 +1,75 @@
+          <ng-container *ngIf="pattern !== ''">
+            <ng-container i18n>To confirm the deletion, enter</ng-container>&nbsp;
+            <kbd>{{ pattern }}</kbd>&nbsp;
+            <ng-container i18n>and click on</ng-container>&nbsp;
```
s/click on/click/

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html:
```diff
@@ -0,0 +1,60 @@
+          novalidate>
+      <div class="modal-body">
+        <p>
+          <ng-container i18n>To move</ng-container>
```
Please apply the following patch to fix the text issue.

![auswahl_002](https://user-images.githubusercontent.com/1897962/45826114-0d8a2700-bcf4-11e8-97c3-3f5c91be5486.png)

```
--- src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html	(date 1533458219000)
+++ src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-move-modal/rbd-trash-move-modal.component.html	(date 1537454024000)
@@ -10,9 +10,9 @@
           novalidate>
       <div class="modal-body">
         <p>
-          <ng-container i18n>To move</ng-container>
+          <ng-container i18n>To move</ng-container>&nbsp;
           <kbd>{{ poolName }}/{{ imageName }}</kbd>&nbsp;
-          <ng-container i18n>to trash, click</ng-container>
+          <ng-container i18n>to trash, click</ng-container>&nbsp;
           <kbd i18n>Move Image</kbd>.&nbsp;
           <ng-container i18n>Optionally, you can pick a different expiration date.</ng-container>
         </p>
```

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html:
```diff
@@ -0,0 +1,53 @@
+          novalidate>
+      <div class="modal-body">
+        <p>
+          <ng-container i18n>To restore</ng-container>
```
Please apply the following patch to fix the text display issue.

![auswahl_003](https://user-images.githubusercontent.com/1897962/45826061-f0edef00-bcf3-11e8-87ce-11073064d131.png)

```
--- src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html	(date 1533458219000)
+++ src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-trash-restore-modal/rbd-trash-restore-modal.component.html	(date 1537454340000)
@@ -10,9 +10,9 @@
           novalidate>
       <div class="modal-body">
         <p>
-          <ng-container i18n>To restore</ng-container>
+          <ng-container i18n>To restore</ng-container>&nbsp;
           <kbd>{{ poolName }}/{{ imageName }}@{{ imageId }}</kbd>,&nbsp;
-          <ng-container i18n>type the image's new name and click</ng-container>
+          <ng-container i18n>type the image's new name and click</ng-container>&nbsp;
           <kbd i18n>Restore Image</kbd>.
         </p>
```

votdev commented:
The date chooser dialog is displayed erroneous (Chromium Version 69.0.3497.81).

![peek 2018-09-20 16-43](https://user-images.githubusercontent.com/1897962/45826288-85585180-bcf4-11e8-843d-3ad01e994d33.gif)

votdev commented:
Is it somehow possible to adapt the generic deletion dialog behaviour? I know the dialog has some more features if the selected image has an expiration date, so you can not reuse the deletion modal dialog.

![auswahl_004](https://user-images.githubusercontent.com/1897962/45826916-ae2d1680-bcf5-11e8-8de1-650a2700d4c7.png)

tspmelo commented:
@votdev Did you restarted the `ng serve`? This PR modifies `angular.json` and that is not automatically read.

I will look into the deletion modal.

votdev commented:
> @votdev Did you restarted the `ng serve`? This PR modifies `angular.json` and that is not automatically read.
> 
> I will look into the deletion modal.

After running ``npm install`` it works, but I don't know if this was the reason.

tspmelo commented:
Just pushed some changes and the PR is now ready to review and merge.

- I added a temporary fix for the datepicker so we don't have to wait for ngx-bootstrap to release a new version.
- I'm now using the `DeletionModalComponent` and have removed the `TrashDeleteModalComponent`
- Using the new `TableActionsComponent`

I have also updated the screenshots.

callithea commented:
jenkins test dashboard

callithea commented:
QA run failed: http://pulpito.ceph.com/laura-2018-09-24_15:30:44-rados:mgr-wip-lpaduano-testing-23351-distro-basic-smithi/

callithea commented:
jenkins test dashboard

tspmelo commented:
jenkins retest this please

tspmelo commented:
jenkins retest this please

callithea commented:
QA run passed: http://pulpito.ceph.com/laura-2018-09-26_08:54:02-rados:mgr-wip-lpaduano-testing-23351-2-distro-basic-smithi/
